models/base.py

```diff
@@ -487,35 +565,305 @@ def _get_fullpath_single(self, download=True, ext=None, nofile=None):
+        # origmd5 holds the md5sum (hashlib.hash object) of the original file,
+        #   *unless* the original file is already the right file in the local file store
+        #   (in which case it's None)
+        # localpath holds the absolute path of where the file should be written in the local file store
+        relpath = pathlib.Path( self.filepath if extension is None else self.filepath + extension )
+        localpath = pathlib.Path( self.local_path ) / relpath
+        if isinstance( data, bytes ):
+            path = "passed data"
+        else:
+            if isinstance( data, str ):
+                path = pathlib.Path( data )
+            elif isinstance( data, pathlib.Path ):
+                path = data
+            else:
+                raise TypeError( f"data must be bytes, str, or Path, not {type(data)}" )
+            path = path.absolute()
+            data = None
+
+        alreadyinplace = False
+        mustwrite = False
+        origmd5 = None
+        if not localpath.exists():
+            mustwrite = True
+        else:
+            if not localpath.is_file():
+                raise RuntimeError( f"{localpath} exists but is not a file!  Can't save." )
+            if localpath == path:
+                alreadyinplace = True
+                _logger.debug( f"FileOnDiskMixin.save: local file store path and original path are the same: {path}" )
+            else:
+                if ( not overwrite ) and ( not exists_ok ):
+                    raise FileExistsError( f"{localpath} already exists, cannot save." )
+                if verify_md5:
+                    origmd5 = hashlib.md5()
+                    if data is None:
+                        with open( path, "rb" ) as ifp:
+                            data = ifp.read()
+                    origmd5.update( data )
+                    localmd5 = hashlib.md5()
+                    with open( localpath, "rb" ) as ifp:
+                        localmd5.update( ifp.read() )
+                    if localmd5.hex_digest() != origmd5.hex_digest():
+                        if overwrite:
+                            _loger.debug( f"Existing {localpath} md5sum mismatch; overwriting." )
+                            mustwrite = True
+                        else:
+                            raise ValueError( f"{localpath} exists, but its md5sum {localmd5.hex_digest()} does not "
+                                              f"match md5sum of {path} {origmd5.hex_digest()}" )
+                else:
+                    if overwrite:
+                        _logger.debug( f"Overwriting {localpath}" )
+                        mustwrite = True
+                    elif exists_ok:
+                        _logger.debug( f"{localpath} already exists, not verifying md5 nor overwriting" )
+                    else:
+                        # raise FileExistsError( f"{localpath} already exists, not saving" )
```

Comment:
If you look at the commet on the next line, this is actually a gratuitous else; logically, the code flow should never get here, since it won't get to the containing else block if either overwrite or exists_ok is already True.  I will sometimes do this (put in as "This should never happen" RuntimeError for code flow that should never happen, but when coding is an obvious think to think might happen.  Two reasons; first, as a flag when reading the code that I thought about this, and second, so if something gets edited earlier that violates the assumptions that keep the "This should never happen" from happening, there will be an exception if it does then in fact later happen.

Reply:
I agree. I just wanted to get rid of the commented line. The rest of the else clause seems ok to me.